File: Computer Programming/C++/Hashing/Prefix Sum/Subarray Sum Equals K.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) 
    {
        unordered_map<int,int> mp;
        int n = nums.size();
        vector<int>prefix(n);
        prefix[0] = nums[0];
        for(int i = 1;i < n; i++)
        {
            prefix[i] = prefix[i-1] + nums[i];
        }    

        mp[0] = 1;
        int count = 0;
        for(int i = 0;i<n;i++)
        {
            int diff = prefix[i] - k;

            if(mp.count(diff))
            {
                count = count + mp[diff];
            }

            mp[prefix[i]] ++;
        }

        return count;
    }
};

File: Computer Programming/Python/Hashing/Subarray Sum Equals K.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        mp = {}

        prefix = {}
        n = len(nums)
        prefix[0] = nums[0]
        for i in range(1,n):
            prefix[i] = prefix[i-1] + nums[i] 
        
        mp[0] = 1
        count = 0
        for i in range(n):
            diff = prefix[i] - k

            if diff in mp:
                count += mp[diff]

            if prefix[i] in mp:
                mp[prefix[i]] += 1
            else:
                mp[prefix[i]] = 1
        
        return count
